### PR TITLE
Product details page

### PR DESCRIPTION
**Pull request by Yuliia143**

### CODE REVIEW

File: src/app/pages/checkout/checkout.component.ts [hunk 1]
```diff
@@ -39,7 +39,8 @@ export class CheckoutComponent implements OnInit, OnDestroy {
         promo: new FormControl(''),
         mailing: new FormControl(false, [Validators.requiredTrue]),
         terms: new FormControl(false, [Validators.requiredTrue]),
-        goods: new FormArray([], [Validators.required])
+        goods: new FormArray([], [Validators.required]),
+        property: new FormControl('Psc')
```
Comment (iyarymovych-reenbit) [inline src/app/pages/checkout/checkout.component.ts]: What does `'Psc'` mean?
Please, create semantic variable for this default value for `property` form control.

File: src/app/pages/checkout/checkout.component.ts [hunk 2]
```diff
@@ -53,10 +54,10 @@ export class CheckoutComponent implements OnInit, OnDestroy {
     }
 
     private patchGoods(): void {
-        const goodsFormArray = this.checkoutForm.get('goods') as FormArray;
```
Comment (iyarymovych-reenbit) [inline src/app/pages/checkout/checkout.component.ts]: Why have you changed it back?

File: src/app/pages/checkout/checkout.component.ts [hunk 3]
```diff
@@ -100,10 +101,10 @@ export class CheckoutComponent implements OnInit, OnDestroy {
 
     public handleCountValue(event: Event, good: GoodInterface): void {
         const target = event.target as HTMLInputElement;
-        if (target.valueAsNumber < 1) {
```
Comment (iyarymovych-reenbit) [inline src/app/pages/checkout/checkout.component.ts]: It looks like you have resolved merge conflicts incorrectly.

File: src/app/pages/product/product.component.ts [hunk 4]
```diff
@@ -8,18 +8,33 @@ import { ProductInterface } from '@shared/interfaces/product-interface';
...
     }
 
     private initializeProduct(): void {
         this.route.data.subscribe(data => {
             this.product = data.product;
         });
     }
+
+    private initializeProducts(): void {
+        this.route.data.subscribe(data => {
```
Comment (iyarymovych-reenbit) [inline src/app/pages/product/product.component.ts:32]: Do unsubscribers from all subscriptions.

File: src/app/pages/product/tabs/tab/tab.component.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<div [hidden]="!active" class="tab">
```
Comment (Vsumskyi) [inline src/app/pages/product/tabs/tab/tab.component.html]: In my opinion, you should use *ngIf

File: src/app/pages/product/tabs/tabs.component.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    ngAfterContentInit(): void {
+        const activeTabs = this.tabs.filter((tab) => tab.active);
+        if (activeTabs.length === 0) {
+            this.selectTab(this.tabs.first);
+        }
+    }
+
+    selectTab(tab): void {
```
Comment (Vsumskyi) [inline src/app/pages/product/tabs/tabs.component.ts]: public or private?

File: src/app/shared/services/cart.service.ts [hunk 7]
```diff
@@ -23,14 +23,14 @@ export class CartService {
         return this.goods.find(item => item.good.id === good.id);
     }
 
-    public addToCart(good: ProductInterface): GoodInterface[] {
+    public addToCart(good: ProductInterface, count: number = 1): GoodInterface[] {
         const existedGood = this.isExistGoodInCart(good);
         if (!existedGood) {
-            this.goods.push({ good, count: 1 });
+            this.goods.unshift({ good, count });
```
Comment (Vsumskyi) [inline src/app/shared/services/cart.service.ts]: I dont like this 'unshift', with this method you have a issue with performance.
Maybe you should use 'push' and apply 'flex-direction: column-reverse' in yours styles

File: src/app/app-routing.module.ts [hunk 8]
```diff
@@ -21,7 +21,8 @@ const routes: Routes = [
         path: 'products/:id',
         loadChildren: () => import('./pages/product/product.module').then(m => m.ProductModule),
         resolve: {
-            product: ProductResolver
+            product: ProductResolver,
+            products: ProductsResolver
```
Comment (Vsumskyi) [inline src/app/app-routing.module.ts:25]: Can do mistake very easy with this naming...

**Review by Vsumskyi — APPROVED** [review]
Approved

File: src/app/pages/product/product-details/product-details.component.html [hunk 9]
```diff
@@ -1,35 +1,115 @@
...
-                <li>Delivery area: Czech republic</li>
-            </ul>
+    <div class="product__details">
+        <h1 class="product__label">{{product.title | titlecase}}</h1>
+        <div class="product__rating">
+            <img *ngFor="let i of stars"
+                 [src]="i <= product.rating? 'assets/images/svg/ic-star-black.svg' : 'assets/images/svg/ic-star-empty.svg'"
+                 alt="star">
+            <span class="product__reviews" *ngIf="product.reviews.length">({{product.reviews.length}}
+                customer review)</span>
```
Comment (maksym-syniuk) [inline src/app/pages/product/product-details/product-details.component.html]:
```suggestion
                ({{product.reviews.length}} customer review)
              </span>
```
(Bad indentation)

File: src/app/pages/product/product-details/product-details.component.html [hunk 10]
```diff
@@ -1,35 +1,115 @@
...
         </div>
-        <div>
-            <p>36.23 USD</p>
-            <p>1 pcs</p>
-            <button class="green-button" (click)="handleAddToCart(product)">Add to cart</button>
+        <div class="product__description">{{product.description.main}}</div>
+        <ul class="product__list">
+            <li class="product__item">
+                <span class="product__item__label">County:</span>
```
Comment (maksym-syniuk) [inline src/app/pages/product/product-details/product-details.component.html]:
```suggestion
                <span class="product__item__label">Country:</span>
```

File: src/app/pages/product/product-details/product-details.component.html [hunk 11]
```diff
@@ -1,35 +1,115 @@
...
+                    <div class="tab__block">
+                        <h3 class="tab__title">Origins</h3>
+                        <p class="tab__content">{{product.description.origins}}</p>
+                    </div>
+                    <div class="tab__block">
+                        <h3 class="tab__title">How to cook</h3>
+                        <p class="tab__content">{{product.description.howToCook}}</p>
+                    </div>
+                </app-tab>
+                <app-tab title="Reviews" [count]="product.reviews.length">
```
Comment (maksym-syniuk) [inline src/app/pages/product/product-details/product-details.component.html]:
```suggestion
                <app-tab title="Reviews" [count]="product.reviews?.length">
```
If there aren't any reviews you will get an error in the way you wrote

File: src/app/pages/product/product-details/product-details.component.html [hunk 12]
```diff
@@ -1,35 +1,115 @@
...
+                        <li class="tab__block" *ngFor="let review of product.reviews">
+                            <div class="tab__review">
+                                <span class="tab__author">{{review.author | titlecase}}</span>
+                                <span class="tab__date">{{review.date | date}}</span>
+                            </div>
+                            <p class="tab__content">{{review.body}}</p>
+                        </li>
+                    </ul>
+                </app-tab>
+                <app-tab title="Questions" [count]="product.reviews.length">
```
Comment (maksym-syniuk) [inline src/app/pages/product/product-details/product-details.component.html]:
```suggestion
                <app-tab title="Questions" [count]="product.questions?.length">
```

File: src/app/pages/product/recommended-products/recommended-products.component.html [hunk 13]
```diff
@@ -1,9 +1,18 @@
...
+    <div class="recommended__title">
+        <h2>You will maybe love</h2>
+        <button class="recommended__button" (click)="getNextRecommended()"
+                [disabled]="this.endIndex >= this.productsFromCategory.length">
+            <span>More products</span>
+            <img src="assets/images/svg/ic-arrow-right-green.svg" alt="arrow">
+        </button>
+    </div>
+    <ul class="recommended__list">
+        <li class="recommended__item-empty" *ngIf="!this.productsFromCategory.length">
```
Comment (maksym-syniuk) [inline src/app/pages/product/recommended-products/recommended-products.component.html]: The same here, you will get an error like this: 
> _Can't read property `length` of `undefined`_

 if there you will receive from Back-End or somewhere else value like `null` or `undefined`



```suggestion
        <li class="recommended__item-empty" *ngIf="!this.productsFromCategory?.length">
```

File: src/app/pages/product/recommended-products/recommended-products.component.ts [hunk 14]
```diff
@@ -1,15 +1,44 @@
...
-  templateUrl: './recommended-products.component.html',
-  styleUrls: ['./recommended-products.component.scss']
+    selector: 'app-recommended-products',
+    templateUrl: './recommended-products.component.html',
+    styleUrls: ['./recommended-products.component.scss']
 })
 export class RecommendedProductsComponent implements OnInit {
+    @Input() products: ProductInterface[];
+    @Input() product: ProductInterface;
+    public productsFromCategory;
```
Comment (maksym-syniuk) [inline src/app/pages/product/recommended-products/recommended-products.component.ts]: type?

File: src/app/pages/product/recommended-products/recommended-products.component.ts [hunk 15]
```diff
@@ -1,15 +1,44 @@
...
+        this.recommendedProducts = this.getRecommendedProducts();
+    }
+
+    private getFromCurrentCategory(): ProductInterface[] {
+        return this.products.filter(product => {
+            if (product.id !== this.product.id && product.category === this.product.category) {
+                return product;
+            }
+        });
+    }
```
Comment (maksym-syniuk) [inline src/app/pages/product/recommended-products/recommended-products.component.ts:29]: You can simplify this method like this, can't you?
```ts
 private getFromCurrentCategory(): ProductInterface[] {
        return this.products.filter(product =>
            product.id !== this.product.id && product.category === this.product.category
        );
    }
```

File: src/app/pages/product/tabs/tabs.component.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+})
+export class TabsComponent implements OnInit, AfterContentInit {
+    @ContentChildren(TabComponent) tabs: QueryList<TabComponent>;
+
+    ngOnInit(): void {
+    }
+
+    ngAfterContentInit(): void {
+        const activeTabs = this.tabs.filter((tab) => tab.active);
+        if (activeTabs.length === 0) {
```
Comment (maksym-syniuk) [inline src/app/pages/product/tabs/tabs.component.ts]:
```suggestion
        if (!activeTabs?.length) {
```

File: src/app/shared/services/cart.service.ts [hunk 17]
```diff
@@ -23,14 +23,14 @@ export class CartService {
...
+    public addToCart(good: ProductInterface, count: number = 1): GoodInterface[] {
         const existedGood = this.isExistGoodInCart(good);
         if (!existedGood) {
-            this.goods.push({ good, count: 1 });
+            this.goods.push({ good, count });
             localStorage.setItem('goods', JSON.stringify(this.goods));
             return this.goods;
         }
-        existedGood.count += 1;
+        count > 1 ? existedGood.count += count : existedGood.count += 1;
```
Comment (maksym-syniuk) [inline src/app/shared/services/cart.service.ts]:
```suggestion
        existedGood.count += count > 1 ? count : 1;
```
(Probably it will work :) )

**Review by maksym-syniuk — APPROVED** [review]

**Pull request merged**